Add MindsDB dbt adapter (#1507)
dbt-labs/docs.getdbt.com: website/docs/reference/warehouse-profiles/mindsdb-profile.md

pip install dbt-mindsdb
```

## Configurations

Basic `profile.yml` for connecting to MindsDB:

```yml
mindsdb:
  outputs:
    dev:
      database: 'mindsdb'
      host: '127.0.0.1'
      password: ''
      port: 47335
      schema: 'mindsdb'
      type: mindsdb
      username: 'mindsdb'
  target: dev

```
| Key      | Required | Description                                          | Example                        |
| -------- | -------- | ---------------------------------------------------- | ------------------------------ |
| type     |    ✔️   | The specific adapter to use                          | `mindsdb`                      |
| host     |    ✔️   | The MindsDB (hostname) to connect to                 | `cloud.mindsdb.com`            |
| port     |    ✔️   | The port to use                                      | `3306`  or `47335`             |
| schema   |    ✔️   | Specify the schema (database) to build models into   | The MindsDB datasource         |
| username |    ✔️   | The username to use to connect to the server         | `mindsdb` or mindsdb cloud user|
| password |    ✔️   | The password to use for authenticating to the server | `pass                          |

> ## Usage

Reviewer: can you add this information to a separate config page as described [here](https://docs.getdbt.com/docs/contributing/documenting-a-new-adapter#topics-and-pages-to-cover)?

- Create dbt project, choose mindsdb database and set up connection
```    
    dbt init <project_name>
```
- To create predictor add dbt model with "predictor" materialization: 
Name of the model is used as name of predictor.
Parameters:
  - integration - name of used integration to get data from and save result to.

>     In has to be created in mindsdb beforehand

Reviewer: how do you create the integration? can you at least link to a page that shows how to do it?

  - predict - field for prediction
  - predict_alias [optional] - alias for predicted field
  - using [optional] - options for configure trained model
```    
    {{
        config(
            materialized='predictor',
            integration='photorep',
            predict='name',
            predict_alias='name',
            using={
                'encoders.location.module': 'CategoricalAutoEncoder',
                'encoders.rental_price.module': 'NumericEncoder'
            }
        )
    }}
      select * from stores
```

- To apply predictor add dbt model with "table" materialization. 
It creates or replaces table in selected integration with results of predictor.
Name of the model is used as name of the table to store prediction results. 

> If you need to specify schema you can do it with dot separator: schema_name.table_name.sql    

Reviewer: can the user not also configure the schema by using a `schema` parameter in `config`? If yes, then please say so. If not, then I'd like to know why!

Author: mindsdb can have several integrations and several schemas (in one dbt project). If we choose to set up the schema in the config we can have only one schema.

> - Create dbt project, choose mindsdb database and set up connection
> ```    
>     dbt init <project_name>
> ```
> - To create predictor add dbt model with "predictor" materialization: 
> Name of the model is used as name of predictor.
> Parameters:
>   - integration - name of used integration to get data from and save result to.
>     In has to be created in mindsdb beforehand
>   - predict - field for prediction
>   - predict_alias [optional] - alias for predicted field
>   - using [optional] - options for configure trained model
> ```    
>     {{
>         config(
>             materialized='predictor',
>             integration='photorep',
>             predict='name',
>             predict_alias='name',
>             using={
>                 'encoders.location.module': 'CategoricalAutoEncoder',
>                 'encoders.rental_price.module': 'NumericEncoder'
>             }
>         )
>     }}
>       select * from stores
> ```
> 
> - To apply predictor add dbt model with "table" materialization. 
> It creates or replaces table in selected integration with results of predictor.
> Name of the model is used as name of the table to store prediction results. 
> If you need to specify schema you can do it with dot separator: schema_name.table_name.sql    
> Parameters:
>   - predictor_name - name of using predictor.
>     It has to be created in mindsdb
>   - integration - name of used integration to get data from and save result to.
>     In has to be created in mindsdb beforehand
> ```    
>     {{ config(materialized='table', predictor_name='TEST_PREDICTOR_NAME', integration='int1') }}
>         select a, bc from ddd where name > latest
> ```

Reviewer: ```suggestion
Create dbt project, choose mindsdb as the database and set up the connection. Verify your connection works `dbt debug`

`dbt init <project_name>`

To create a predictor, create a dbt model with a "predictor" materialization. The name of the model will be the name of predictor.

#### Parameters:
- `integration` - name of used integration to get data from and save result to. Must be created in mindsdb beforehand
- `predict` - field for prediction
- `predict_alias` [optional] - alias for predicted field
- `using` [optional] - options for configure trained model

```sql
-- my_first_model.sql    
    {{
        config(
            materialized='predictor',
            integration='photorep',
            predict='name',
            predict_alias='name',
            using={
                'encoders.location.module': 'CategoricalAutoEncoder',
                'encoders.rental_price.module': 'NumericEncoder'
            }
        )
    }}
      select * from stores
```

To apply predictor add dbt model with "table" materialization. It creates or replaces table in selected integration with results of predictor. The models name will be the name of the resulting predictions table. If you need to specify schema you can do it with dot separator: `schema_name.table_name.sql`
    
#### Parameters:
- `predictor_name` - name of using predictor.
    It has to be created in mindsdb
- `integration` - name of used integration to get data from and save result to (must be created in mindsdb beforehand)

```sql
-- my_models_predictions.sql 
    {{ config(materialized='table', predictor_name='TEST_PREDICTOR_NAME', integration='int1') }}
        select a, bc from ddd where name > latest
```
```

Reviewer: this suggestion above got a little kooky, but the major changes I made (beyond edits for clarity) were:
- no nested bullets, only bullets for the parameters
- adding `sql` to the beginning of of codeblocks
- adding a comment with an example file name for the sql

## Testing

- Install dev requirements
```    
  pip install -r dev_requirements.txt
```
- Run pytest
```    
  python -m pytest tests/

> ```

Reviewer: not sure this section is relevant for end-users, probably makes sense to keep in the GitHub repo's `README.md` or, better yet, a `CONTRIBUTING.md`.
```suggestion
## Testing

- Install dev requirements
```    
  pip install -r dev_requirements.txt
```
- Run pytest
```    
  python -m pytest tests/
```
```

```sql
-- my_first_model.sql    
    {{
        config(
            materialized='predictor',
            integration='photorep',
            predict='name',
            predict_alias='name',
            using={
                'encoders.location.module': 'CategoricalAutoEncoder',
                'encoders.rental_price.module': 'NumericEncoder'
            }
        )
    }}
      select * from stores
```

To apply predictor add dbt model with "table" materialization. It creates or replaces table in selected integration with results of predictor.
Name of the model is used as name of the table to store prediction results.
If you need to specify schema you can do it with dot separator: schema_name.table_name.sql  

#### Parameters
- `predictor_name` - name of the predictor. It has to be created in mindsdb.
- `integration` - name of used integration to get data from and save result to. Must be created in mindsdb beforehand using the [`CREATE DATABASE` syntax](https://docs.mindsdb.com/sql/create/databases/).

```    
    {{ config(materialized='table', predictor_name='TEST_PREDICTOR_NAME', integration='photorep') }}
        select a, bc from ddd where name > latest

> ```

Reviewer: can you move this whole Usage section to `mindsdb-configs.md`? configs is how to use the adapter. profile is how to connect.
